src/execution/sql/table_vector_iterator.cpp

```diff
@@ -1,63 +1,206 @@
 #include "execution/sql/table_vector_iterator.h"
 
-#include <tbb/blocked_range.h>
-#include <tbb/parallel_for.h>
-#include <tbb/task_scheduler_init.h>
-
-#include <memory>
+#include <limits>
+#include <numeric>
+#include <utility>
 #include <vector>
 
 #include "execution/exec/execution_context.h"
+#include "execution/sql/column_vector_iterator.h"
+#include "execution/sql/thread_state_container.h"
 #include "execution/util/timer.h"
+#include "loggers/execution_logger.h"
+#include "tbb/parallel_for.h"
+#include "tbb/task_scheduler_init.h"
 
 namespace terrier::execution::sql {
+
 TableVectorIterator::TableVectorIterator(exec::ExecutionContext *exec_ctx, uint32_t table_oid, uint32_t *col_oids,
                                          uint32_t num_oids)
     : exec_ctx_(exec_ctx), table_oid_(table_oid), col_oids_(col_oids, col_oids + num_oids) {}
 
-TableVectorIterator::~TableVectorIterator() {
-  exec_ctx_->GetMemoryPool()->Deallocate(buffer_, projected_columns_->Size());
-}
+TableVectorIterator::~TableVectorIterator() = default;
 
 bool TableVectorIterator::Init() {
-  // Find the table
+  // No-op if already initialized
+  if (IsInitialized()) {
+    return true;
+  }
+
+  // Set up the table and the iterator.
   table_ = exec_ctx_->GetAccessor()->GetTable(table_oid_);
   TERRIER_ASSERT(table_ != nullptr, "Table must exist!!");
+  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->begin());
+  const auto &table_col_map = table_->GetColumnMap();
+
+  // Configure the vector projection, create the column iterators.
+  std::vector<storage::col_id_t> col_ids;
+  std::vector<TypeId> col_types(col_oids_.size());
+  column_iterators_.reserve(col_oids_.size());
+  for (uint64_t idx = 0; idx < col_oids_.size(); idx++) {
+    auto col_oid = col_oids_[idx];
+    auto col_type = GetTypeId(table_col_map.at(col_oid).col_type_);
+    auto storage_col_id = table_col_map.at(col_oid).col_id_;
 
-  // Initialize the projected column
-  TERRIER_ASSERT(!col_oids_.empty(), "There must be at least one col oid!");
-  auto pc_init = table_->InitializerForProjectedColumns(col_oids_, common::Constants::K_DEFAULT_VECTOR_SIZE);
-  buffer_ = exec_ctx_->GetMemoryPool()->AllocateAligned(pc_init.ProjectedColumnsSize(), alignof(uint64_t), false);
-  projected_columns_ = pc_init.Initialize(buffer_);
+    col_ids.emplace_back(storage_col_id);
+    col_types[idx] = col_type;
+    column_iterators_.emplace_back(GetTypeIdSize(col_type));
+  }
+
+  // Create an owning vector.
+  vector_projection_.SetStorageColIds(col_ids);
+  vector_projection_.Initialize(col_types);
+  vector_projection_.Reset(common::Constants::K_DEFAULT_VECTOR_SIZE);
+
+  // All good.
   initialized_ = true;
+  return true;
+}
 
-  // Begin iterating
-  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->begin());
+bool TableVectorIterator::Init(uint32_t block_start, uint32_t block_end) {
+  // TODO(WAN): code duplication. Though it does make it pretty clear.
```

Comment:
I don't agree. The only difference is the one line that creates the iterator - the rest is gnarly OID to index mapping work. I say have one `Init(begin,end)` that uses `begin==end` to indicate a full scan. Or, keep the two `Init()` functions, but dispatch to a common `InitCommon(begin,end)` (or something similar) that uses `table_->begin()` if `begin==end`, and uses the boundaries otherwise?

Reply:
Going with begin=0, end=u32::max.